core/src/main/java/hudson/ProxyConfiguration.java

```diff
@@ -259,6 +259,40 @@ public PasswordAuthentication getPasswordAuthentication() {
         return con;
     }
 
+    /**
+     * This method should be used wherever {@link URL#openConnection()} to internet URLs is invoked directly and a redirection is possible.
+     * @since TODO
+     */
+    public static URLConnection openAllowsRedirection(URL url) throws IOException {
+        URLConnection con = ProxyConfiguration.open(url);
+        // A redirection from http to https (or vise versa) returns a 302 response status. Force redirection
+        int responseCode = HttpURLConnection.HTTP_OK;
+        // TODO responseCode always will be OK for non-http connections. Processing needed
+        if (con instanceof HttpURLConnection) {
+            // prevent problems from misbehaving plugins disabling redirects by default
+            HttpURLConnection httpCon = ((HttpURLConnection) con);
+            httpCon.setInstanceFollowRedirects(true);
+            responseCode = httpCon.getResponseCode();
+            if (HttpURLConnection.HTTP_OK != responseCode && HttpURLConnection.HTTP_MOVED_TEMP != responseCode && HttpURLConnection.HTTP_MOVED_PERM != responseCode && HttpURLConnection.HTTP_SEE_OTHER != responseCode) {
+                // No OK Status and no redirection
+                throw new IOException("Could not open " + url);
+            } else if (HttpURLConnection.HTTP_OK != responseCode) {
```

Comment:
This is too lenient, accepting non-redirect responses. What if it's HTTP 201 Created, which also uses the Location header?

Reply:
If we get a 201 response code, the exception will be thrown. I'm not sure if it makes sense to allow another 2xx code furthermore status 200.

Reply:
@daniel-beck any extra feedback about this point, please? I am not sure if I totally get you.